File: next/plugins/configstore/configstore.py
"""C1: viewer.config-store — plugin-level configuration CRUD (framework §10.2).

Configuration here means *plugin-level* config (affects all instances, e.g.
the chat plugin's roles/agents list) — deliberately distinct from instance
state (C2). RPC surface: `config:_:get/set/list`. Every mutation republishes
the plugin's FULL config to the `config:{plugin}:config` mailbox (state =
full value, framework §5.6). Storage: one atomic-write JSON file.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import logging
import os
import signal
from pathlib import Path
from typing import Any

from sdk import Ctx, Plugin, slot

logger = logging.getLogger(__name__)

MANIFEST: dict[str, Any] = {
    "id": "config-store",
    "version": "0.1.0",
    "slots": {"get": {}, "set": {}, "list": {}},
    "emits": {"config": {}},
}


def mailbox_channel(plugin_id: str) -> str:
    return f"config:{plugin_id}:config"


class ConfigStorePlugin(Plugin):
    manifest = MANIFEST

    def __init__(self, db_path: Path | None = None, **client_kwargs: Any) -> None:
        super().__init__(**client_kwargs)
        self.db_path = db_path or (Path.home() / ".view" / "config-store.json")
        self._data: dict[str, dict[str, Any]] = {}

    async def on_start(self) -> None:
        if self.db_path.is_file():
            self._data = json.loads(self.db_path.read_text())
            logger.info("loaded %d plugin configs from %s", len(self._data), self.db_path)

    def _save(self) -> None:
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self.db_path.with_suffix(".tmp")
        tmp.write_text(json.dumps(self._data, ensure_ascii=False, indent=2))
        tmp.replace(self.db_path)

    @slot("config:_:get")
    async def get(self, ctx: Ctx) -> None:
        plugin_id, key = _plugin_and_key(ctx)
        if plugin_id is None:
            await ctx.respond_error("invalid_request", "missing required field: plugin")
            return
        config = self._data.get(plugin_id, {})
        await ctx.respond(config.get(key) if key else config)

    @slot("config:_:set")
    async def set_(self, ctx: Ctx) -> None:
        plugin_id, key = _plugin_and_key(ctx)
        if plugin_id is None or key is None:
            await ctx.respond_error("invalid_request", "missing required fields: plugin, key")
            return
        value = ctx.value.get("value")
        config = self._data.setdefault(plugin_id, {})
        if value is None:
            config.pop(key, None)
        else:
            config[key] = value
        self._save()
        # State channel: full-value replace, consumers never need an RPC (§5.6).
        assert self.client is not None
        await self.client.set(mailbox_channel(plugin_id), dict(config))
        await ctx.respond({"plugin": plugin_id, "key": key, "value": value})

    @slot("config:_:list")
    async def list_(self, ctx: Ctx) -> None:
        await ctx.respond(dict(self._data))

    def run(self) -> None:
        parser = argparse.ArgumentParser(description="Viewer C1 config-store plugin")
        parser.add_argument("--kernel-ws", required=True)
        parser.add_argument("--db", default=None, help="config JSON file path")
        args = parser.parse_args()
        logging.basicConfig(
            level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s: %(message)s"
        )
        if args.db is not None:
            self.db_path = Path(args.db)
        managed = os.environ.get("VIEWER_MANAGED") == "1"

        async def main() -> None:
            await self.start(args.kernel_ws, managed=managed)
            stopped = asyncio.Event()
            loop = asyncio.get_running_loop()
            for signum in (signal.SIGINT, signal.SIGTERM):
                loop.add_signal_handler(signum, stopped.set)
            await stopped.wait()
            await self.stop()

        asyncio.run(main())


def _plugin_and_key(ctx: Ctx) -> tuple[str | None, str | None]:
    if not isinstance(ctx.value, dict):
        return None, None
    return ctx.value.get("plugin"), ctx.value.get("key")

File: internal/plugins/configstore/configstore.go
// Package configstore implements plugin-level configuration persistence.
package configstore

import (
	"context"
	"log/slog"
	"sync"

	"viewer/internal/plugins/pluginrpc"
	"viewer/internal/plugins/storefile"
	"viewer/sdk/go/busclient"
)

const defaultFilename = "config-store.json"

var Manifest = busclient.Manifest{
	ID: "config-store", Version: "0.1.0",
	Slots: map[string]any{"get": map[string]any{}, "set": map[string]any{}, "list": map[string]any{}},
	Emits: map[string]any{"config": map[string]any{}},
}

type Plugin struct {
	dbPath string
	mu     sync.Mutex
	data   map[string]map[string]any
	client *busclient.Client
}

func New(dbPath string) (*Plugin, error) {
	if dbPath == "" {
		var err error
		dbPath, err = storefile.DefaultPath(defaultFilename)
		if err != nil {
			return nil, err
		}
	}
	return &Plugin{dbPath: dbPath, data: make(map[string]map[string]any)}, nil
}

func MailboxChannel(pluginID string) string { return "config:" + pluginID + ":config" }

func (p *Plugin) Start(ctx context.Context, kernelWS string, managed bool) error {
	if err := storefile.Load(p.dbPath, &p.data); err != nil {
		return err
	}
	if p.data == nil {
		p.data = make(map[string]map[string]any)
	}
	client := busclient.New(kernelWS, Manifest, busclient.WithManaged(managed))
	for pattern, handler := range map[string]func(busclient.Frame){
		"config:_:get":  p.get,
		"config:_:set":  p.set,
		"config:_:list": p.list,
	} {
		if _, err := client.Subscribe(pattern, handler); err != nil {
			_ = client.Close()
			return err
		}
	}
	p.client = client
	if err := client.Connect(ctx); err != nil {
		p.client = nil
		_ = client.Close()
		return err
	}
	slog.Info("config-store started", "db", p.dbPath)
	return nil
}

func (p *Plugin) Close() error {
	if p.client == nil {
		return nil
	}
	return p.client.Close()
}

func (p *Plugin) get(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, valid := stringField(value, "plugin")
	if !ok || !valid {
		p.replyError(frame, "invalid_request", "missing required field: plugin")
		return
	}
	p.mu.Lock()
	config := cloneObject(p.data[pluginID])
	key, hasKey := optionalString(value, "key")
	var result any = config
	if hasKey && key != "" {
		result = config[key]
	}
	p.mu.Unlock()
	p.reply(frame, result)
}

func (p *Plugin) set(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	value, ok := pluginrpc.Object(frame)
	pluginID, validPlugin := stringField(value, "plugin")
	key, validKey := stringField(value, "key")
	if !ok || !validPlugin || !validKey {
		p.replyError(frame, "invalid_request", "missing required fields: plugin, key")
		return
	}
	newValue := value["value"]
	p.mu.Lock()
	config := p.data[pluginID]
	if config == nil {
		config = make(map[string]any)
		p.data[pluginID] = config
	}
	if newValue == nil {
		delete(config, key)
	} else {
		config[key] = newValue
	}
	if err := storefile.Save(p.dbPath, p.data); err != nil {
		p.mu.Unlock()
		p.replyError(frame, "write_error", err.Error())
		return
	}
	full := cloneObject(config)
	if err := p.client.Set(context.Background(), MailboxChannel(pluginID), full); err != nil {
		p.mu.Unlock()
		slog.Error("publish config mailbox", "error", err)
		return
	}
	p.mu.Unlock()
	p.reply(frame, map[string]any{"plugin": pluginID, "key": key, "value": newValue})
}

func (p *Plugin) list(frame busclient.Frame) {
	if pluginrpc.Cancelled(frame) {
		return
	}
	p.mu.Lock()
	result := make(map[string]any, len(p.data))
	for pluginID, config := range p.data {
		result[pluginID] = cloneObject(config)
	}
	p.mu.Unlock()
	p.reply(frame, result)
}

func (p *Plugin) reply(frame busclient.Frame, result any) {
	if err := pluginrpc.Respond(p.client, frame, result); err != nil {
		slog.Error("config-store RPC response failed", "error", err)
	}
}

func (p *Plugin) replyError(frame busclient.Frame, code, message string) {
	if err := pluginrpc.RespondError(p.client, frame, code, message); err != nil {
		slog.Error("config-store RPC error response failed", "error", err)
	}
}

func stringField(value map[string]any, key string) (string, bool) {
	if value == nil {
		return "", false
	}
	result, ok := value[key].(string)
	return result, ok
}

func optionalString(value map[string]any, key string) (string, bool) {
	raw, exists := value[key]
	if !exists || raw == nil {
		return "", false
	}
	result, ok := raw.(string)
	return result, ok
}

func cloneObject(value map[string]any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	result := make(map[string]any, len(value))
	for key, item := range value {
		result[key] = item
	}
	return result
}
